### PR TITLE
Refactor dropout functionality

### PR DESCRIPTION
**Pull request by ZongrenZou**

### CODE REVIEW

File: deepxde/callbacks.py [hunk 1]
```diff
@@ -242,6 +242,38 @@ def on_epoch_end(self):
             )
 
 
+class DropoutUncertainty(Callback):
+    def __init__(self, period=1):
```
Comment (lululxvi) [inline deepxde/callbacks.py:246]: period=1000

File: deepxde/callbacks.py [hunk 2]
```diff
@@ -242,6 +242,38 @@ def on_epoch_end(self):
             )
 
 
+class DropoutUncertainty(Callback):
```
Comment (lululxvi) [inline deepxde/callbacks.py:245]: “”“Uncertainty estimation via MC dropout.

Reference: https://arxiv.org/abs/1506.02142

Warning:
    This cannot be used together with other techniques that have different behaviors during training and testing, such as batch normalization.
”“”

File: deepxde/callbacks.py [hunk 3]
```diff
@@ -242,6 +242,38 @@ def on_epoch_end(self):
...
+    """Uncertainty estimation via MC dropout.
+
+    Reference: https://arxiv.org/abs/1506.02142
+
+    Warning:
+    This cannot be used together with other techniques that have different behaviors during training and testing, such as batch normalization.
```
Comment (lululxvi) [inline deepxde/callbacks.py:251]: Extra 4 empty spaces before "This"

File: deepxde/callbacks.py [hunk 4]
```diff
@@ -242,6 +242,38 @@ def on_epoch_end(self):
...
+        self.period = period
+        self.epochs_since_last = 0
+
+    def on_epoch_end(self):
+        self.epochs_since_last += 1
+        if self.epochs_since_last >= self.period:
+            self.epochs_since_last = 0
+            y_preds = []
+            for _ in range(1000):
+                y_pred_test_one, _ = self.model._run(
```
Comment (lululxvi) [inline deepxde/callbacks.py:265]: Add a comment here:
# No need to compute outputs_losses, only outputs are needed.

File: deepxde/callbacks.py [hunk 5]
```diff
@@ -242,6 +242,38 @@ def on_epoch_end(self):
...
+    Warning:
+        This cannot be used together with other techniques that have different behaviors during training and testing, such as batch normalization.
+    """
+
+    def __init__(self, period=1000):
+        super(DropoutUncertainty, self).__init__()
+        self.period = period
+        self.epochs_since_last = 0
+
+    def on_epoch_end(self):
```
Comment (lululxvi) [inline deepxde/callbacks.py:259]: Add
```
def on_train_end(self):
    self.on_epoch_end()
```

File: deepxde/maps/pytorch/fnn.py [hunk 6]
```diff
@@ -25,4 +25,4 @@ def forward(self, x):
         for linear in self.linears[:-1]:
             x = self.activation(linear(x))
         x = self.linears[-1](x)
-        return x
+        return x
```
Comment (lululxvi) [inline deepxde/maps/pytorch/fnn.py:28]: Why this file is modified?

File: deepxde/model.py [hunk 7]
```diff
@@ -271,14 +267,14 @@ def train(
         self.stop_training = False
         self.train_state.set_data_train(*self.data.train_next_batch(self.batch_size))
         self.train_state.set_data_test(*self.data.test())
-        self._test(uncertainty)
+        self._test()
         self.callbacks.on_train_begin()
         if optimizers.is_external_optimizer(self.opt_name):
             self._train_scipy(display_every, uncertainty)
```
Comment (lululxvi) [inline deepxde/model.py:273]: Delete `uncertainty`
